### DISCUSSION TITLE
`UserManager.on_after_...` in v8.0

### DISCUSSION THREAD

**Discussion by davidbrochart** [fastapi-users/fastapi-users#734] [Q&A] [answered]
FastAPI-Users 8.0 uses a `UserManager` class, on which a `on_after_register` can be defined.
- In case the DB has to be updated after registering, I was wondering how this method should take in a `user_db`, should it directly be e.g. a `SQLAlchemyUserDatabase(UserDB, database, users, oauth_accounts)` or can it also accept a dependency, through which we could pass `get_user_db`?
- What if we also have e.g. a GitHub OAuth router, and we must do something different in that case after registering? I'm a bit lost where this should go now.

**Comment by frankie567** [comment, marked as answer]
Hi @davidbrochart!

## Access DB in handler

Since the method lives in the class, you have access to the user database class through `self`; e.g:

```py
class UserManager(BaseUserManager[users.UserCreate, users.UserDB]):
    user_db_model = users.UserDB
    reset_password_token_secret = settings.secret

    async def on_after_register(
        self, user: users.UserDB, request: Optional[Request] = None
    ):
        user.initialized = True
        await self.user_db.update(user)
```

Alternatively, you can use the [`_update` method](https://github.com/fastapi-users/fastapi-users/blob/a82dfc262da0c4d976a7411b765531e5785880dd/fastapi_users/manager.py#L537-L552):

```py
class UserManager(BaseUserManager[users.UserCreate, users.UserDB]):
    user_db_model = users.UserDB
    reset_password_token_secret = settings.secret

    async def on_after_register(
        self, user: users.UserDB, request: Optional[Request] = None
    ):
        updated_user = await self._update(user, {"initialized": True})
```

## `on_after_register` and OAuth

Indeed, the handlers are now the same in the case of standard or OAuth registration. If you need different logic when authenticating with OAuth you can either:

### Check if there is an OAuth account in the registered user (which I recommend):
   
```py
async def on_after_register(
    self, user: users.UserDB, request: Optional[Request] = None
):
    for oauth_account in user.oauth_accounts:
        if oauth_account.oauth_name == "github":
            print("User has registered with GitHub")
```

### Add a custom method in your class and overload `oauth_callback`:

```py
class UserManager(BaseUserManager[users.UserCreate, users.UserDB]):
    user_db_model = users.UserDB
    reset_password_token_secret = settings.secret

    async def oauth_callback(self, oauth_account, request=None):
        user = await super().oauth_callback(oauth_account, request=request)
        await self.on_after_register_oauth(user, oauth_account, request)
        return user

    async def on_after_oauth_callback(
        self, user: users.UserDB, oauth_account, request: Optional[Request] = None
    ):
        print("User has authenticated with OAuth")
```

Notice that in this case, you won't be able to distinguish between a registration and an authentication. Besides, `on_after_register` will still be called if it's a new user. I invite you to [check the code](https://github.com/fastapi-users/fastapi-users/blob/a82dfc262da0c4d976a7411b765531e5785880dd/fastapi_users/manager.py#L159-L209) to understand what's going on inside. 

Hope that'll help :)

**Reply by davidbrochart** [reply]
That helped a lot, thanks so much for the quick answer!
